geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/AlterQueryServiceCommand.java

```diff
@@ -0,0 +1,209 @@
+  static final String SECURITY_NOT_ENABLED_MESSAGE =
+      "Integrated security is not enabled for this distributed system. Updating the method authorizer requires integrated security to be enabled.";
+  public static final String PARTIAL_FAILURE_MESSAGE =
+      "In the event of a partial failure of this command, re-running the command or restarting failing members should restore consistency.";
+
+  @CliCommand(value = COMMAND_NAME, help = COMMAND_HELP)
+  @CliMetaData(
+      interceptor = "org.apache.geode.management.internal.cli.commands.AlterQueryServiceCommand$Interceptor")
+  @ResourceOperation(resource = ResourcePermission.Resource.CLUSTER,
+      operation = ResourcePermission.Operation.MANAGE)
+  public ResultModel execute(
+      @CliOption(key = METHOD_AUTHORIZER_NAME,
+          help = METHOD_AUTHORIZER_NAME_HELP) String methodAuthorizerName,
+      @CliOption(key = AUTHORIZER_PARAMETERS,
+          help = AUTHORIZER_PARAMETERS_HELP) String[] authorizerParameters,
+      @CliOption(key = SINGLE_AUTHORIZER_PARAMETER,
+          help = SINGLE_AUTHORIZER_PARAMETER_HELP) String singleAuthorizerParameter) {
+    ResultModel result;
+
+    QueryConfigService queryConfigService = getQueryConfigService();
+    Set<String> parametersSet = null;
+    if (methodAuthorizerName != null) {
+      if (!isSecurityEnabled()) {
+        return ResultModel.createError(SECURITY_NOT_ENABLED_MESSAGE);
+      }
+      if (Boolean.parseBoolean(
+          System.getProperty(GEMFIRE_PREFIX + "QueryService.allowUntrustedMethodInvocation"))) {
+        return ResultModel.createError(DEPRECATED_PROPERTY_ERROR);
+      }
+      if (authorizerParameters != null) {
+        parametersSet = new HashSet<>(Arrays.asList(authorizerParameters));
+      } else if (singleAuthorizerParameter != null) {
+        parametersSet = new HashSet<>(Collections.singleton(singleAuthorizerParameter));
+      }
+      populateMethodAuthorizer(methodAuthorizerName, parametersSet, queryConfigService);
+    }
+
+    Set<DistributedMember> targetMembers = findMembers(null, null);
+    if (targetMembers.size() > 0) {
+      Object[] args = new Object[] {methodAuthorizerName, parametersSet};
+      List<CliFunctionResult> functionResults =
+          executeAndGetFunctionResult(new AlterQueryServiceFunction(), args,
+              targetMembers);
+      String footer = null;
+      for (CliFunctionResult cliResult : functionResults) {
+        if (!cliResult.isSuccessful()) {
+          footer = PARTIAL_FAILURE_MESSAGE;
+          break;
+        }
+      }
+      result =
+          ResultModel.createMemberStatusResult(functionResults, null, footer, false, true);
+    } else {
+      result = ResultModel.createInfo(NO_MEMBERS_FOUND_MESSAGE);
+    }
+    result.setConfigObject(queryConfigService);
+    return result;
+  }
+
+  boolean isSecurityEnabled() {
+    return ((InternalCache) CacheFactory.getAnyInstance()).getSecurityService()
+        .isIntegratedSecurity();
+  }
```

Comment:
I don't know this part of the code so the question might make no sense at all: here we're checking wether security is enabled for the JVM on which the gfsh command is running, shouldn't we check wether security is enabled at the cluster level (locator) instead?.

Reply:
I am not sure how this methodAuthorizer is related to the securityService, but commands are executed on jmxManager which is usually a locator or a server with jmx-manager enabled (in the case where is no locator involved). When security service is enabled, it should be enabled on all members (locators and servers).

Reply:
If cluster config is enabled then checking the locator is sufficient, but if cluster config is not enabled then it's possible for a locator to no have a security service but the servers that connect to it to have one, leading to incorrect behaviour. This check has been moved to the `AlterQueryServiceFunction` class to ensure correct behaviour.